Add cloud awareness to sunpy

wmoapl opened the pull request:
Added way for `Map` and `TimeSeries` factories to recognize and read in URI file paths. If URI, it will use `fsspec` to open, read, and look for files. Descriptions to changes in code:

`Pyproject.toml`
- added fsspec as package dependency

`Sunpy/io/_filetools.py`
- added way to open URI filetype using fsspec.open to detect_filetype

`Sunpy/map/map_factory.py`
- added logic to separate URI file path from regular filepath and parses URIs using `parse_uri` function

`Sunpy/timeseries/timeseries_factory.py`
- added logic to separate URI file path from regular filepath and parses URIs using `parse_uri` function
- moved out previous code for parsing timeseries results into separate function

`Sunpy/util/io.py`
-  `parse_uri` - reads in a list of fsspec OpenFile objects; follows same structure as `parse_path`
- `is_uri` - recognize URI objects; code taken from pandas (also being used in pyspedas)

wmoapl committed:
initial sunpy fix

wmoapl committed:
added fsspec as main dependency

wmoapl committed:
Merge branch 'main' into htm-s3-awareness

wmoapl commented:
Might want to think about how to handle anonymous keys. Most NASA repos are gonna be public, so it would be good to not require keys in case the user doesn't have an AWS account.

vtoyedens commented on sunpy/timeseries/timeseries_factory.py:
```diff
@@ -338,6 +341,46 @@ def _parse_args(self, *args, **kwargs):
+        if isinstance(r, GenericTimeSeries):
+            return [r]#all_ts += [r]
+        elif isinstance(r, pathlib.Path):
+            return [self._check_registered_widgets(filepath=r, **kwargs)]#all_ts += [self._check_registered_widgets(filepath=r, **kwargs)]
```
Remove the commented code

wmoapl replied:
done

vtoyedens commented on sunpy/timeseries/timeseries_factory.py:
```diff
@@ -338,6 +341,46 @@ def _parse_args(self, *args, **kwargs):
                 raise type(e)(msg) from e
         return all_ts
 
+    def _parse_ts_results(self, r, **kwargs):
```
Is this just a code refactor because it was messy? I don't see any changes regarding uris

wmoapl replied:
Yes, just refactoring. It use to be code under `_parse_path(self, path, **kwargs)`. I'm pulling it out because I use the same code for `_parse_uri(self, obj_list, **kwargs)`.

vtoyedens reviewed (commented):
Minor comments, clean implementation. Do tests exist for these files already?

wmoapl committed:
adding fsspec anon option, tests

wmoapl committed:
finalized tests

wmoapl commented:
Added ability to read in Map (using astropy) with anonymous credentials. This allows anyone with or without credentials to read from a public bucket. Use this call:
```
from sunpy.map import Map
Map("s3://gov-nasa-hdrl-data1/sdo/aia/20100513/0131/sdo_aia_h2_20100513T000400_0131_v1.fits", fsspec_kwargs={'anon': True})
```
Unfortunately, still cannot read CDFs anonymously using `timeseries_factory.py`. This would require a change in cdflib to use fsspec instead of boto3. I've changed the code to replicate the functionality in `map_factory.py` in case the `cdflib` change ever comes.

Also added tests of reading from s3 to `map/tests/test_map_factory.py` and `timeseries/tests/test_timeseries_factory.py`. Map factory test will be done with anonymous credentials while the timeseries test requires valid credentials. It took me a while to get tests going. This is what I ended up doing in a clean environment:
```
pip install -e ".[tests]" 
tox -e py312-online 
```
This runs tests for the entire package which is super slow. If anyone figures out how to run it for individual modules let me know....

vtoyedens reviewed (approved):
Looks good to me, ran tests and ran branch code

edmondb commented:
Tested those example code lines you sent, although the one regarding anonymous credentials didn't fully work for me (received a `Killed` message on the EC2 that I'm testing with. Perhaps I needed more memory.

Only recommendation is consistency with the importation of fsspec throughout. Only place I saw a difference was really in 
https://github.com/APL-HTM-S3-AWARE/sunpy/blob/fc887bc8e987e4294fe6f2227be854c678a267ec/sunpy/map/map_factory.py#L6

The others do not use the `fs` alias`. Didn't test the tests either as I also have issues with the actual SunPy pip install regarding the gcc lib. If you want me to keep testing for completeness, I can.

wmoapl committed:
conform fsspec import

edmondb commented:
Looks good now to me! Time to merge...
